service/storage_service.cc

```diff
@@ -5377,6 +5378,96 @@ future<> storage_service::update_fence_version(token_metadata::version_t new_ver
     });
 }
 
+static
+locator::tablet_replica get_leaving_replica(const locator::tablet_info& tinfo, const locator::tablet_transition_info& trinfo) {
+    std::unordered_set<locator::tablet_replica> leaving(tinfo.replicas.begin(), tinfo.replicas.end());
+    for (auto&& r : trinfo.next) {
+        leaving.erase(r);
+    }
+    if (leaving.empty()) {
+        throw std::runtime_error(format("No leaving replicas"));
+    }
+    if (leaving.size() > 1) {
+        throw std::runtime_error(format("More than one leaving replica"));
+    }
+    return *leaving.begin();
+}
+
+inet_address storage_service::host2ip(locator::host_id host) {
+    auto ip = _group0->address_map().find(raft::server_id(host.uuid()));
+    if (!ip) {
+        throw std::runtime_error(::format("Cannot map host {} to ip", host));
+    }
+    return *ip;
+}
+
+// Streams data from a given tablet replica to the pending tablet replica on this node.
+future<> storage_service::stream_tablet(locator::global_tablet_id tablet) {
+    // The coordinator does not execute global token metadata barrier before jumping to "streaming" stage, so we need
```

Comment:
It's better not to, because read barrier here is cheaper and more fault tolerant than the global barrier.

Reply:
`raft_topology_cmd_handler` also has the same code. It also checks for outdated coordinator. Why no call `stream_tablet` through it instead of adding a separate RPC?

Reply:
I considered it, but raft_topology_cmd_handler expects that commands are serialized, and here we want to stream in parallel.

Reply:
Can you elaborate. As far as I see  `handle_tablet_streaming` sends a bunch rpcs to a bunch of nodes and waits for it to be complete. This is more or less what `raft_topology_cmd` does as well. In the case of `handle_tablet_streaming`  you have different params for different host, right? Each param says what should be streamed and where. Is this why separate RPC is needed? If yous then two question: can the node itself figure out this info like we do with regular streaming and if not then can we call `raft_topology_cmd` instead the new one but from iniside `handle_tablet_streaming`?

Reply:
The RPC receiver figures out all the info except for which tablet to stream, and it must be the receiving replica. We want to stream tablets independently, so we have RPC per-tablet and per receiver rather than a single one. In this PR we stream tablets in tandem, but that will change in the next PR.

Regarding raft_topology_cmd reuse, it's answered here: https://github.com/scylladb/scylladb/pull/14601#discussion_r1274979394